src/validator.cpp

```diff
@@ -1635,6 +1708,530 @@ void Validator::ValidatorImpl::validateMathMLElements(const XmlNodePtr &node, co
+                                                    const XmlNodePtr &node,
+                                                    const ComponentPtr &component)
+{
+    if (mathmlChildCount(parentNode) != 3) {
+        addMathMLIssue("Math has a '" + node->name() + "' element without two MathML siblings.",
+                       Issue::ReferenceRule::MATH_MATHML,
+                       component);
+
+        return false;
+    }
+
+    return true;
+}
+
+bool Validator::ValidatorImpl::hasAtLeastTwoMathmlSiblings(const XmlNodePtr &parentNode,
+                                                           const XmlNodePtr &node,
+                                                           const ComponentPtr &component)
+{
+    if (mathmlChildCount(parentNode) < 3) {
+        addMathMLIssue("Math has a '" + node->name() + "' element without at least two MathML siblings.",
+                       Issue::ReferenceRule::MATH_MATHML,
+                       component);
+
+        return false;
+    }
+
+    return true;
+}
+
+size_t Validator::ValidatorImpl::hasOneOrTwoMathmlSiblings(const XmlNodePtr &parentNode,
+                                                           const XmlNodePtr &node,
+                                                           const ComponentPtr &component)
+{
+    auto childCount = mathmlChildCount(parentNode);
+
+    if ((childCount != 2) && (childCount != 3)) {
+        addMathMLIssue("Math has a '" + node->name() + "' element without one or two MathML siblings.",
+                       Issue::ReferenceRule::MATH_MATHML,
+                       component);
+
+        return 0;
+    }
+
+    return childCount - 1;
+}
+
+bool Validator::ValidatorImpl::isFirstMathmlChild(const XmlNodePtr &parentNode,
+                                                  const XmlNodePtr &node,
+                                                  const ComponentPtr &component)
+{
+    if (!mathmlChildNode(parentNode, 0)->equals(node)) {
+        addMathMLIssue("Math has a '" + node->name() + "' element which is not the first MathML sibling.",
+                       Issue::ReferenceRule::MATH_MATHML,
+                       component);
+
+        return false;
+    }
+
+    return true;
+}
+
+bool Validator::ValidatorImpl::isSecondMathmlChild(const XmlNodePtr &parentNode,
+                                                   const XmlNodePtr &node,
+                                                   const ComponentPtr &component)
+{
+    if (!mathmlChildNode(parentNode, 1)->equals(node)) {
+        addMathMLIssue("Math has a '" + node->name() + "' element which is not the second MathML sibling.",
+                       Issue::ReferenceRule::MATH_MATHML,
+                       component);
+
+        return false;
+    }
+
+    return true;
+}
+
+bool Validator::ValidatorImpl::hasFirstMathmlSiblingWithName(const XmlNodePtr &parentNode,
+                                                             const XmlNodePtr &node,
```

Comment:
I called it that way because it's checking for the first MathML sibling of `node`. To do that check, we need to use `node`'s parent and go through its children and skip `node` itself.

Reply:
yeah, I'm understanding this collection of methods now...its not so much about how the checks are being performed but how they are used.